### PR TITLE
Management API library - Service API

### PR DESCRIPTION
**Pull request by vadmeste**
Just a first attempt of a code that works but probably needs some discussion

### CODE REVIEW

File: pkg/madmin/utils.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+// Verify if input endpoint URL is valid.
+func isValidEndpointURL(endpointURL string) error {
+	if endpointURL == "" {
+		return ErrInvalidArgument("Endpoint url cannot be empty.")
+	}
+	url, err := url.Parse(endpointURL)
+	if err != nil {
+		return ErrInvalidArgument("Endpoint url cannot have fully qualified paths.")
```
Comment (krisis) [inline pkg/madmin/utils.go]: Isn't this error different from the one below? This means the URL was malformed and not parseable.

**Review by krisis — COMMENTED** [review]
Initial comments.

File: pkg/madmin/utils.go (hunk 1 above)
Comment (vadmeste) [reply pkg/madmin/utils.go]: > Isn't this error different from the one below? This means the URL was malformed and not parseable.

Agree

**Comment by vadmeste** [comment]
@harshavardhana, issues were addressed, but I am thinking to add a new package under minio-go/pkg/utils which has some useful functions like isValidEndpointURL, isValidDomain, isValidIP etc, .. what do you think ?

**Comment by harshavardhana** [comment]
> @harshavardhana, issues were addressed, but I am thinking to add a new package under minio-go/pkg/utils which has some useful functions like isValidEndpointURL, isValidDomain, isValidIP etc, .. what do you think ?

@vadmeste if we can make it generic enough but as a thumb rule naming `utils` as a package can cause lot of unrelated functions to be dumped there which is a bad practice for a package itself -  `urlutil` is a better word for this, like httputil package. Deals with anything URL related.

**Comment by harshavardhana** [comment]
Is this still a WIP @vadmeste ?

**Comment by vadmeste** [comment]
> Is this still a WIP @vadmeste ?

No, I've updated the title.

File: pkg/madmin/client_test.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ */
+
+// Package madmin_test
+package madmin_test
+
+import "testing"
+
+func TestMAdminClient(t *testing.T) {
+	/* cl, _ := madmin.NewAdminClient("localhost:9000", "minio", "minio123", false)
```
Comment (harshavardhana) [inline pkg/madmin/client_test.go]: Can we remove the comment?

File: pkg/madmin/service-api.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		Type BackendType
+		// Following fields are only meaningful if BackendType is XL.
+		OnlineDisks  int // Online disks during server startup.
+		OfflineDisks int // Offline disks during server startup.
+		ReadQuorum   int // Minimum disks required for successful read operations.
+		WriteQuorum  int // Minimum disks required for successful write operations.
+	}
+}
+
+// ServiceStatus - Call Service Status API
```
Comment (harshavardhana) [inline pkg/madmin/service-api.go]: Little more descriptive comment describing the purpose.

File: pkg/madmin/service-api.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+	err = json.Unmarshal(respBytes, &storageInfo)
+	if err != nil {
+		return ServiceStatusMetadata{}, err
+	}
+
+	return storageInfo, nil
+}
+
+// ServiceStop - Call Service Stop API
```
Comment (harshavardhana) [inline pkg/madmin/service-api.go]: Little more descriptive comment describing the purpose.

File: pkg/madmin/service-api.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return errors.New("Got " + resp.Status)
+	}
+
+	return nil
+}
+
+// ServiceRestart - Call Service Restart API
```
Comment (harshavardhana) [inline pkg/madmin/service-api.go]: Little more descriptive comment describing the purpose.

File: pkg/madmin/utils.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// getEndpointURL - construct a new endpoint.
+func getEndpointURL(endpoint string, secure bool) (*url.URL, error) {
+	if strings.Contains(endpoint, ":") {
+		host, _, err := net.SplitHostPort(endpoint)
+		if err != nil {
+			return nil, err
+		}
+		if !isValidIP(host) && !isValidDomain(host) {
```
Comment (harshavardhana) [inline pkg/madmin/utils.go]: You can use s3utils here..

File: pkg/madmin/requests.go [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+	accessKeyID string
+	// SecretAccessKey required for authorized requests.
+	secretAccessKey string
+
+	// User supplied.
+	appInfo struct {
+		appName    string
+		appVersion string
+	}
+	endpointURL string
```
Comment (harshavardhana) [inline pkg/madmin/requests.go]: You can use endpointURL to be url.URL here not string. Makes validation easier.

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Everything else looks good.

**Comment by vadmeste** [comment]
All issues are addressed @harshavardhana, also I vendorized s3utils package.

**Comment by vadmeste** [comment]
All issues are addressed @harshavardhana, also I vendorized s3utils and s3signer packages.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**